File: 1035-cousins-in-binary-tree/1035-cousins-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isCousins(TreeNode* root, int x, int y) {
        if(root==NULL) return false;
        queue<TreeNode*>q;
        q.push(root);
        while(!q.empty()){
            int n=q.size();
            bool xExist=false; bool yExist=false;
            while(n--){
                TreeNode* curr=q.front(); q.pop();
                if(curr->val==x) xExist=true;
                if(curr->val==y) yExist=true;
                if(curr->left && curr->right){
                    if((curr->left->val==x && curr->right->val==y) || 
                    (curr->left->val==y && curr->right->val==x)){
                        return false;
                    }
                }
                if(curr->left) q.push(curr->left);
                if(curr->right) q.push(curr->right);
            }
            if(xExist && yExist) return true;
        }
        return false;
    }
};

File: 1035-cousins-in-binary-tree/1035-cousins-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int xDepth,yDepth;
    TreeNode xParent,yParent;
    private void helper(TreeNode root,TreeNode parent, int level, int x, int y){
        if(root==null) return ;
        if(root.val==x){
            xDepth=level;
            xParent=parent;
        }
        if(root.val==y){
            yDepth=level;
            yParent=parent;
        }
        helper(root.left,root,level+1,x,y);
        helper(root.right,root,level+1,x,y);


    }
    public boolean isCousins(TreeNode root, int x, int y) {
        helper(root,null,0,x,y);
        return xDepth==yDepth && xParent!=yParent;
    }
}